### ISSUE TITLE
macOS 10.12, SickRage unable to find unrar in $PATH

### ISSUE THREAD

**Issue by adlibdalom** [SickRage/SickRage#2592]
### Before submitting your issue:

Enable debug logging in SickRage settings, reproduce the error (be sure to disable after the bug is fixed)

**Branch/Commit**: master/d15b7003f8663b76d01fcbf0566389b2f31e0564
**OS**: macOS 10.12 (16B2657)
**What you did**: Manual (or automatic) Post-Processing. I've tried updating unrar, reinstalled it with brew and making sure it's in $PATH, but to no avail.
**What happened**: Nothing (apart from the log below)
**What you expected**: All of my downloaded episodes to start unraring and eventually be sorted.
**Logs**:
```
2016-11-04 22:21:07 Thread-16 :: [d15b700] Failed Unrar archive *****.rar: No suitable RAR unpacker installed
2016-11-04 22:21:06 Thread-16 :: [d15b700] Failed Unrar archive *****.rar: No suitable RAR unpacker installed
2016-11-04 22:21:05 Thread-16 :: [d15b700] Failed Unrar archive *****.rar: No suitable RAR unpacker installed
2016-11-04 22:21:05 Thread-16 :: [d15b700] Failed Unrar archive *****.rar: No suitable RAR unpacker installed
2016-11-04 22:21:04 Thread-16 :: [d15b700] Failed Unrar archive *****.rar: No suitable RAR unpacker installed
2016-11-04 22:21:03 Thread-16 :: [d15b700] Failed Unrar archive *****.rar: No suitable RAR unpacker installed
```

**Comment by SickRage2** [comment]
Thanks for the issue report! Before a real human comes by, please make sure your report has all the below criteria checked

- [ ] Include basic information: Branch/Commit, OS, What you did, What happened, What you expected
- [ ] Enable debug logging (be sure to disable after the bug is fixed)
- [ ] Post debug logs, either inline (for smaller logs) or using gist

Please make sure you also read [how to create an issue](https://github.com/SickRage/sickrage-issues/blob/master/README.md) and followed all of the steps. 

The title should describe your issue. Having "SR not working" or "I get this bug" for 100 issues, isn't really helpful. We will close issues if there isn't enough information.

Sometimes the devs may seem like grunts and respond with short answers. This isn't (always) because the dev hates you, but because he's on mobile or busy fixing bugs. If something isn't clear, please let us know, and this bot may get updated to automatically answer you.

Thanks!

**Comment by bobbysteel** [comment]
This happened to me recently. It seems to be something errant with the PATH setup in Brew. For me it worked by downloading the unrar binary, copying fresh to /usr/local/bin/unrar, symlinking /usr/bin/unrar to /usr/local/bin/unrar then checking permissions set properly to 0755 or whatever you need them.

**Comment by adlibdalom** [comment]
@bobbysteel wrote:
> symlinking /usr/bin/unrar to /usr/local/bin/unrar

The problem with this approach is that it's not allowed with SIP running, and I don't really think SickRage should need to bypass SIP just to get `unrar` to work.

**Comment by narimanshariat** [comment]
I'm having the same issues and have tried a number of different things. One major issue seems to be that in macOS 10.12 /usr/bin/* is read-only. So no symlinks or copying files to that path.

I can access unrar from the command line and everything seems to be working properly except SickRage having access to it which is the real problem.

**Comment by arkdelkaos** [comment]
I'm trying to edit the path from `/usr/bin` to `/usr/local/bin` at the code, but I can't find where SickRage locate the unrar binary on MacOS. May someone help me find where I should change it?  
I don't want to bypass SIP just for SR to work again (never had problems before) with RAR files; it's kinda overkill, iykwim ;)

**Comment by OmgImAlexis** [comment]
Guys why not just edit your $PATH and add the path to unrar in there? That'd make a lot more sense than symlinking and changing permissions.

**Comment by narimanshariat** [comment]
Already tried that.

`living-room:~ livingroom$ echo $PATH
/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Applications/Server.app/Contents/ServerRoot/usr/bin:/Applications/Server.app/Contents/ServerRoot/usr/sbin`

first path is `/usr/local/bin` where unrar is located. Again it is accessible from the command line but SR doesn't have access to see it.

**Comment by bobbysteel** [comment]
Does changing the path manually in /lib/rarfile/__init__.py fix it?

**Comment by narimanshariat** [comment]
I changed this line
https://github.com/SickRage/SickRage/blob/master/lib/rarfile/__init__.py#L170
to 
`UNRAR_TOOL = "/usr/local/bin/unrar"`
restarted SR and still no luck.

Log:
`2016-12-06 08:45:00 WARNING Thread-15 :: Unpacking Not Supported, disabling unpack setting 2016-12-06 08:45:00 ERROR Thread-15 :: [a1bf3b3] Rar Not Supported: No suitable RAR unpacker installed AAUnpackerNotInstalled: No suitable RAR unpacker installed AA raise UnpackerNotInstalled("No suitable RAR unpacker installed") AA File "/Applications/SickRage/lib/unrar2/unix.py", line 50, in call_unrar AA return call_unrar([cmd]+soptions+['--',self.archiveName]+files) AA File "/Applications/SickRage/lib/unrar2/unix.py", line 121, in call AA stdoutdata, stderrdata = self.call('v', []).communicate() AA File "/Applications/SickRage/lib/unrar2/unix.py", line 67, in init AA RarFileImplementation.init(self, password) AA File "/Applications/SickRage/lib/unrar2/__init__.py", line 100, in __init__ AA testing = RarFile(rar_path).read_files('*test.txt') AA File "/Applications/SickRage/sickbeard/webserve.py", line 4389, in isRarSupported AATraceback (most recent call last): 2016-12-06 08:44:51 ERROR Thread-15 :: [a1bf3b3] Rar Not Supported: No suitable RAR unpacker installed`

**Comment by bobbysteel** [comment]
Weird. How about adding the full path in this file: /lib/unrar2/unix.py
I had the same issue but adding it in my $PATH fixed it (albeit on 10.11 ages ago)

**Comment by bobbysteel** [comment]
Is execute permission set on your /usr/bin/local/unrar?

**Comment by narimanshariat** [comment]
I reverted the previous file (init.py)

Then I updated
`https://github.com/SickRage/SickRage/blob/master/lib/unrar2/unix.py#L42`
to: `for command in ('/usr/local/bin/unrar', 'rar'):` and success. I don't get the error message anymore.

**Comment by narimanshariat** [comment]
Thanks @bobbysteel. I wonder if an option can be added in the config to allow users to set the path manually.

**Comment by bobbysteel** [comment]
Definitely seems like a PATH issue on your machine. I'm confused why it wouldn't be detecting it normally though in your PATH as I tested it on my machine with a file in /usr/bin/ and not in /usr/local/bin and Python runs it without issue.

Perhaps try updating /etc/paths.d like this: https://www.cyberciti.biz/faq/appleosx-bash-unix-change-set-path-environment-variable/ then rebooting and trying that again?

**Comment by OmgImAlexis** [comment]
@miigotu any ideas?

**Comment by narimanshariat** [comment]
Here are some results of your suggestions.
`living-room:~ livingroom$ ls -al /usr/local/bin/unrar 
lrwxr-xr-x  1 livingroom  staff  31 Dec  1 01:40 /usr/local/bin/unrar -> ../Cellar/unrar/5.4.5/bin/unrar`

`living-room:~ livingroom$ python -c "import os; print os.environ['PATH']"
/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Applications/Server.app/Contents/ServerRoot/usr/bin:/Applications/Server.app/Contents/ServerRoot/usr/sbin`

**Comment by miigotu** [comment]
Fairly sure you can just put unrar in the SR dir itself also. That lib should probably be replaced with a maintained version. I have had an open PR for an unrelated bug in that lib for over a year:
https://github.com/kyegupov/py-unrar2/pulls

**Comment by arkdelkaos** [comment]
````
python -c "import os; print os.environ['PATH']"
/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin
````
@bobbysteel...but no luck.
I'm gona try @narimanshariat 's solution next :)

**Comment by miigotu** [comment]
@arkdelkaos the path is set separate for each user with SIP. You are checking as your user I assume. It all depends how SR is started, either with administrator privileges or not (not sure what script you are using to start it).

**Comment by miigotu** [comment]
You could try the rarfile branch. A setting can now be coded in to add the unrar path to the settings also if necessary.

`rarfile.UNRAR_TOOL = '/blah/unrar'`

**Comment by arkdelkaos** [comment]
@miigotu
````
ps aux | grep sickrage
Ark              34308   0,0  0,0  2423384    228 s004  R+    6:54PM   0:00.00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn sickrage
````
I think I'm executing it *myself*

**Comment by bobbysteel** [comment]
@miigotu's recommendation to put the unrar in the SR directory is a good one as well. Have you tried that? 
Go into your main SickRage folder then
    `sudo ln -s /usr/local/bin/unrar .`

**Comment by arkdelkaos** [comment]
Putting unrar into the SR directory, doesn't work...but editing `lib/unrar2/unix.py`as @narimanshariat said, works perfectly :)  
btw, @miigotu, I'm executing SR with my user

**Comment by adlibdalom** [comment]
I invested some time into this to get to the bottom of the actual cause of SickRage not being able to locate `unrar` despite it definitely being in `$PATH`.
From what I gathered it seems that it's generally a bad idea to have SickRage run as a `LaunchAgent` or `LaunchDaemon` in the `/Library` context. As soon as I changed to a user-context `~/Library/LaunchAgent`, things started to work drastically better.
What I guess it that, `$PATH` (more specifically the user added specifics of `$PATH` (i. e. `/usr/local/bin/`) are added firstly when a user logs on. Before that, `$PATH` is really just the most basic paths like `/usr/bin`, `/bin`, `/usr/sbin` and `/sbin`.
Now, everything works for me without having to symlink anything in `SIP` protected paths or such.

Also, I started using the `brew` version of `python` (2.7.12 as of this writing) when running SickRage through the `LaunchAgent`, as that seems to also magically aid with finding the correct (`brew`) version of `OpenSSL` to make SickRage able to verify certificates correctly.
